Broken compatibility with OpenSSL 1.0.2k

### Description

CPR v1.11.2 introduced compile error on Centos7:
```
FAILED: cpr/CMakeFiles/cpr.dir/ssl_ctx.cpp.o 
/usr/local/bin/clang++ -DCPR_ENABLE_SSL -DCPR_USE_SYSTEM_CURL -DCURL_STATICLIB -DOPENSSL_BACKEND_USED -I/[REDACTED]/dep_cpr-prefix/src/dep_cpr/include -I/[REDACTED]/dep_cpr-prefix/src/dep_cpr-build/cpr_generated_includes -isystem /[REDACTED]/curl-install/include -fPIC -stdlib=libc++ -Wall -Wextra -Wpedantic -Werror -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-nonportable-system-include-path -Wno-exit-time-destructors -Wno-undef -Wno-global-constructors -Wno-switch-enum -Wno-old-style-cast -Wno-covered-switch-default -Wno-undefined-func-template -O3 -DNDEBUG -std=gnu++17 -MD -MT cpr/CMakeFiles/cpr.dir/ssl_ctx.cpp.o -MF cpr/CMakeFiles/cpr.dir/ssl_ctx.cpp.o.d -o cpr/CMakeFiles/cpr.dir/ssl_ctx.cpp.o -c /[REDACTED]/cpr/dep_cpr-prefix/src/dep_cpr/cpr/ssl_ctx.cpp
/[REDACTED]/cpr/dep_cpr-prefix/src/dep_cpr/cpr/ssl_ctx.cpp:18:10: fatal error: 'openssl/pemerr.h' file not found
   18 | #include <openssl/pemerr.h>
      |          ^~~~~~~~~~~~~~~~~~
1 error generated.
```
I ask you to return compatibility with old OpenSSL version because it was not so long ago in 1.10.5.

### Example/How to Reproduce

First way (I guess):
- Get OpenSSL 1.0.2k release;
- Build OpenSSL and install somewhere but default location;
- Configure cpr specifying OpenSSL root installation dir (-DOPENSSL_ROOT_DIR=/path/to/installed/openssl102k)
- Build cpr.

Second way (I got):
- Run Centos7 (Docker will suit);
- Install depencencies including system OpenSSL (1.0.2k);
- Configure cpr;
- Build cpr.

### Possible Fix

Did not investigated yet, but cold guess:
- Ifdef for OpenSSL version excluding problem header if version too low (header introduced in 2019, 1.0.2k is too old);
- Avoid header usage even for modern versions of OpenSSL if appropriate;
- Ifdefs for old and modern OpenSSL versions exploiting same functionality.

### Where did you get it from?

GitHub (branch e.g. master)

### Additional Context/Your Environment

- OS: Centos
- Version: 7
- Clang: 19
- C++ standard: 17
- C++ standard library: libc++
- 2-stage build, CPR gets compiled during main project configure stage using ExternalProject CMake feature.
- cURL: 8.13 (built from source one step before cpr)
- cpr: v1.11.2 tag

CMake output:
```
-- The CXX compiler identification is Clang 19.1.7
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/local/bin/clang++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CXX standard: 17
-- C++ Requests CMake Options
-- =======================================================
--   CPR_GENERATE_COVERAGE: OFF
--   CPR_CURL_NOSIGNAL: OFF
--   CURL_VERBOSE_LOGGING: OFF
--   CPR_USE_SYSTEM_GTEST: OFF
--   CPR_USE_SYSTEM_CURL: On
--   CPR_ENABLE_CURL_HTTP_ONLY: Off
--   CPR_ENABLE_SSL: ON
--   CPR_FORCE_OPENSSL_BACKEND: OFF
--   CPR_FORCE_WINSSL_BACKEND: OFF
--   CPR_FORCE_DARWINSSL_BACKEND: OFF
--   CPR_FORCE_MBEDTLS_BACKEND: OFF
--   CPR_ENABLE_LINTING: OFF
--   CPR_ENABLE_CPPCHECK: OFF
--   CPR_BUILD_TESTS: OFF
--   CPR_BUILD_TESTS_SSL: OFF
--   CPR_BUILD_TESTS_PROXY: OFF
--   CPR_BUILD_VERSION_OUTPUT_ONLY: OFF
--   CPR_SKIP_CA_BUNDLE_SEARCH: OFF
--   CPR_USE_BOOST_FILESYSTEM: OFF
--   CPR_DEBUG_SANITIZER_FLAG_THREAD: OFF
--   CPR_DEBUG_SANITIZER_FLAG_ADDR: OFF
--   CPR_DEBUG_SANITIZER_FLAG_LEAK: OFF
--   CPR_DEBUG_SANITIZER_FLAG_UB: OFF
--   CPR_DEBUG_SANITIZER_FLAG_ALL: OFF
-- =======================================================
-- Performing Test THREAD_SANITIZER_AVAILABLE
-- Performing Test THREAD_SANITIZER_AVAILABLE - Success
-- Performing Test ADDRESS_SANITIZER_AVAILABLE
-- Performing Test ADDRESS_SANITIZER_AVAILABLE - Success
-- Performing Test LEAK_SANITIZER_AVAILABLE
-- Performing Test LEAK_SANITIZER_AVAILABLE - Success
-- Performing Test UNDEFINED_BEHAVIOUR_SANITIZER_AVAILABLE
-- Performing Test UNDEFINED_BEHAVIOUR_SANITIZER_AVAILABLE - Success
-- Performing Test ALL_SANITIZERS_AVAILABLE
-- Performing Test ALL_SANITIZERS_AVAILABLE - Success
-- Automatically detecting SSL backend.
-- Detecting SSL backend...
-- Found OpenSSL: /usr/lib64/libcrypto.so (found version "1.0.2k")
-- SSL auto detect: Using OpenSSL.
-- Found CURL: /[REDACTED]/curl-install/lib64/cmake/CURL/CURLConfig.cmake (found version "8.13.0-DEV") found components: HTTP HTTPS
-- Curl 8.13.0-DEV found on this system.
-- Configuring done (2.9s)
-- Generating done (0.0s)
```

--- a/.gitignore
+++ b/.gitignore
@@ -61,3 +61,6 @@ compile_commands.json
 
 # macOS
 .DS_Store
+
+# CLion
+.idea/

--- a/cpr/ssl_ctx.cpp
+++ b/cpr/ssl_ctx.cpp
@@ -15,19 +15,23 @@
 #include <openssl/bio.h>
 #include <openssl/err.h>
 #include <openssl/pem.h>
-#include <openssl/pemerr.h>
 #include <openssl/ssl.h>
 #include <openssl/x509.h>
 #include <openssl/x509_vfy.h>
 
-// openssl/types.h was added in later version of openssl and is therefore not always available.
-// This is for example the case on Ubuntu 20.04.
+// openssl/types.h was added in later version of openssl (starting from 3.0.0) and is therefore not always available.
+// This is for example the case on Ubuntu 20.04 or Centos 7.
 // We try to include it if available to satisfy clang-tidy.
 // Ref: https://github.com/openssl/openssl/commit/50cd4768c6b89c757645f28519236bb989216f8d
-#if __has_include(<openssl/types.h>)
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
 #include <openssl/types.h>
 #else
 #include <openssl/ossl_typ.h>
+#endif
+
+// openssl/pemerr.h was added in 1.1.1a
+#if OPENSSL_VERSION_NUMBER >= 0x1010101fL
+#include <openssl/pemerr.h>
 #endif
 
 namespace cpr {

--- a/cpr/ssl_ctx.cpp
+++ b/cpr/ssl_ctx.cpp
@@ -14,6 +14,7 @@
 
 #include <openssl/bio.h>
 #include <openssl/err.h>
+#include <openssl/opensslv.h>
 #include <openssl/pem.h>
 #include <openssl/ssl.h>
 #include <openssl/x509.h>

